src/System.Security.Permissions/tests/PrincipalPermissionTests.cs

```diff
@@ -0,0 +1,329 @@
+            string className = "System.Security.Permissions.PrincipalPermission, ";
+            Assert.NotNull(p);
+            Assert.True(!p.IsUnrestricted());
+            PrincipalPermission copy = (PrincipalPermission)p.Copy();
+            Assert.Equal(p.IsUnrestricted(), copy.IsUnrestricted());
+            SecurityElement se = p.ToXml();
+            Assert.True((se.Attributes["class"] as string).StartsWith(className));
+            Assert.Equal("1", (se.Attributes["version"] as string));
+        }
+
+        [Fact]
+        public void PermissionStateUnrestricted()
+        {
+            PrincipalPermission p = new PrincipalPermission(PermissionState.Unrestricted);
+            Assert.NotNull(p);
+            Assert.True(p.IsUnrestricted());
+            PrincipalPermission copy = (PrincipalPermission)p.Copy();
+            Assert.Equal(p.IsUnrestricted(), copy.IsUnrestricted());
+            // Note: Unrestricted isn't shown in XML
+        }
+
+        [Fact]
+        public void Name()
+        {
+            PrincipalPermission p = new PrincipalPermission("user", null);
+            Assert.True(!p.IsUnrestricted());
+        }
+
+        [Fact]
+        public void UnauthenticatedName()
+        {
+            PrincipalPermission p = new PrincipalPermission("user", null, false);
+            Assert.True(!p.IsUnrestricted());
+        }
+
+        [Fact]
+        public void Role()
+        {
+            PrincipalPermission p = new PrincipalPermission(null, "users");
+            Assert.True(!p.IsUnrestricted());
+        }
+
+        [Fact]
+        public void UnauthenticatedRole()
+        {
+            PrincipalPermission p = new PrincipalPermission(null, "users", false);
+            Assert.True(!p.IsUnrestricted());
+        }
+
+        [Fact]
+        public void NameRole()
+        {
+            PrincipalPermission p = new PrincipalPermission("user", "users", true);
+            Assert.True(!p.IsUnrestricted());
+        }
+
+        [Fact]
+        public void UnauthenticatedNameRole()
+        {
+            PrincipalPermission p = new PrincipalPermission("user", "users", false);
+            Assert.True(!p.IsUnrestricted());
+        }
+
+        [Fact]
+        public void AuthenticatedNullNull()
+        {
+            PrincipalPermission p = new PrincipalPermission(null, null, true);
+            Assert.True(p.IsUnrestricted());
+        }
+
+        [Fact]
+        public void FromXmlNull()
+        {
+            PrincipalPermission p = new PrincipalPermission(PermissionState.None);
+            Assert.Throws<ArgumentNullException>(() => p.FromXml(null));
+        }
+
+        [Fact]
+        public void FromXmlInvalidPermission()
+        {
+            PrincipalPermission p = new PrincipalPermission(PermissionState.None);
+            SecurityElement se = p.ToXml();
+            // can't modify - so we create our own
+            SecurityElement se2 = new SecurityElement("IInvalidPermission", se.Text);
```

Comment:
Is the tag name what makes this fail?  I'm concerned that the next test wouldn't fail if the "2" became a "1".